### ISSUE TITLE
NavigationView regaining focus throws exception in v11-preview6

### ISSUE THREAD

**Issue by kurtyoder** [amwx/FluentAvalonia#324]
**Describe the bug**
My app has multiple tab pages. Tab Page 2 contains a NavigationView. If the user selects a menu item from the NavigationView and then navigates to Tab Page 1, then returns to Tab Page 2 and clicks on a menu item again, the app crashes with a invalid cast exception. 

> System.InvalidCastException: Unable to cast object of type '[ProjectName].ViewModels.Orders.OpenOrderViewModel' to type '[ProjectName].ViewModels.DashboardBaseViewModel'.
>    at CompiledAvaloniaXaml.XamlIlHelpers.[ProjectName].ViewModels.DashboardBaseViewModel,[ProjectName].Categories!Getter(Object)
>    at Avalonia.Data.Core.ClrPropertyInfo.Get(Object target) in /_/src/Avalonia.Base/Data/Core/ClrPropertyInfo.cs:line 27
>    at Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings.InpcPropertyAccessor.get_Value() in /_/src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/CompiledBindings/PropertyInfoAccessorFactory.cs:line 88
>    at Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings.InpcPropertyAccessor.SendCurrentValue() in /_/src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/CompiledBindings/PropertyInfoAccessorFactory.cs:line 132

_OpenOrderViewModel_ hosts the current navigation view model which inherits from _DashboardBaseViewModel_. So not sure why it is trying to make that cast?

**This code worked in version 2.0.0-preview4.**

_OpenOrderView_

 ```
<controls:NavigationView     
        MenuItemsSource="{Binding Categories}"
        Content="{Binding NavigationCurrentViewModel}"
        SelectedItem="{Binding SelectedCategory}">

        <controls:NavigationView.MenuItemTemplate>
         ...
        </controls:NavigationView.MenuItemTemplate>
 </controls:NavigationView>
```
_OpenOrderViewModel_

```
[Reactive] public DashboardBaseViewModel NavigationCurrentViewModel { get; set; }

public IconNavigationCategory? SelectedCategory
{
    get => _selectedCategory;
    set
    {
        this.RaiseAndSetIfChanged(ref _selectedCategory, value);
        if (SelectedCategory != null)
            SetCurrentPage();
    }
}

```

**Screenshots**
_Selecting a menu item selects a DashboardBaseViewModel_
![image](https://user-images.githubusercontent.com/11448589/227745707-cdd7d5a7-7566-4a84-8dc1-c127a02a12c7.png)

_Returning from Tab Page 1 and clicking on any of the menu items will crash the app_
![image](https://user-images.githubusercontent.com/11448589/227745715-b7a7277e-be1c-41d7-b0f4-6fb7f522f6e9.png)

**Desktop/Platform (please complete the following information):**
 - OS:  (11/10)
 - FluentAvalonia Version [2.0.0-preview6]
 - Avalonia Version [11.0.0-preview6]

**Additional context**
When the user moves to Tab Page 1, the _SelectedCategory_ is set to null so clearly the NavigationView is losing focus. When the user returns, no menu item is selected as shown in picture 2. 

Since _NavigationCurrentViewModel_ is not directly linked to _SelectedCategory,_ the view model is still being displayed when the user navigates back, even though the NavigationView doesn't have a selected item. I obviously want this behavior, ideally the NavigationView wouldn't lose selection either...

I am calling this a bug since my code worked before, but perhaps I am doing this the wrong way.

**Comment by amwx** [comment]
You say it worked in preview4, did it work in preview5? 

The invalid cast exception is coming from the CompiledBindings code upstream, can you try turning off CompiledBindings and see if it functions normally? 

The NavigationView is removed from the visual tree if you change the selected tab (which is how the TabView works), so yes it is losing focus. Although, there isn't anything in the NavigationView code that modifies selection when unloading. Do you have a minimal repro I can use to look into this.

**Comment by kurtyoder** [comment]
I tested it with preview5 and the same exception was thrown.

After turning off compile bindings for the view that is the content of the NavigationView it functions as expected, so I'll just run it that way until it gets fixed or roll back to preview4.

In order to get the exception in the minimal repo, I had to add WhenActivated logic to _DashboardBaseViewModel_ (In the repo the logic is in _TabViewModelBase_). So, I think the issue may lie along those lines? 

Here is the repo https://github.com/kurtyoder/NavigationViewMinimal/tree/master/NavigationViewMinimal.Avalonia 

I also dug up two other exceptions, while I was testing that may be related the same problem. I have detailed those in the readme file of the repo. 

Let me know if you need me to get/try anything else.

**Comment by amwx** [comment]
Thanks for the info and repro. It might be somewhere in the ReactiveUI - compiled binding exchange. I'll take a closer look sometime this week.

**Comment by amwx** [comment]
Ok, I've finally had a chance to look at this. I think I have a fix for the NavigationView not remaining selected - that's related to the ItemsSource being cleared when its removed from the tree and when the NavView was added we didn't have the `Loaded` event where WinUI attempts to reselect the item. That will be in the next update.

As for the crash - it is CompiledBinding related. Why it worked before and not now, I have no idea. But, what's happening is you set the x:DataType on the page to `tabPage:HasNavigationViewModel`, and bind the NavView's Content in as `{Binding NavigationCurrentViewModel}`. As a result, the CompiledBinding stuff expects that content to be of type `HasNavigationViewModel` and when its not, you're getting the `InvalidCastException`.

I was able to get it to work by setting the `ContentTemplate`:
```xaml
<controls:NavigationView.ContentTemplate>
	<DataTemplate DataType="viewModels:ViewModelBase">
		<ContentPresenter Content="{Binding }" />
	</DataTemplate>
</controls:NavigationView.ContentTemplate>
```

I don't know what can be done about this. I don't think there's anything I can do on my end and that this is just a flaw of CompiledBindings. I'd suggest to either use the ContentTemplate workaround or stick with a ReflectionBinding for the Content.

CompiledBindings also support casting, you might be able to try casting the content binding, if this works, its probably the better solution:
`Content="{Binding ((viewModels:TabViewModelBase)NavigationCurrentViewModel)}"`
